client/gutenberg/editor/edit-post/components/header/header-toolbar/index.js

```diff
@@ -23,30 +29,110 @@ import {
 	NavigableToolbar,
 } from '@wordpress/editor';
 
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import Drafts from 'layout/masterbar/drafts';
+import Site from 'blocks/site';
+import { addSiteFragment } from 'lib/route';
+import { recordTracksEvent } from 'state/analytics/actions';
+import { getSelectedSite } from 'state/ui/selectors';
+import { getRouteHistory } from 'state/ui/action-log/selectors';
+
+function getCloseButtonPath( routeHistory, site, post ) {
+	// @see post-editor/editor-ground-control/index.jsx
+	const lastNonEditorPath = findLast(
+		routeHistory,
+		action => ! action.path.match( /^\/gutenberg\/(post|page|(edit\/[^\/]+))\/[^\/]+(\/\d+)?$/i )
```

Comment:
I wonder if we should explicitly set something in global state, or pass through a url in the query param, for example  what we'd normally do for a login redirect. It'd be clearer to update closing logic in that case, if we wanted to do something custom for a particular entry point.